### ISSUE TITLE
Macro lets LLVM do more work than it should

### ISSUE THREAD

**Issue by Ten0** [Lucretiel/lazy_format#9]
I'm not sure what the use of:
https://github.com/Lucretiel/lazy_format/blob/e4a5c49f7aecc44306b790996155645e7dbfb395/src/lib.rs#L51

is: it seems that mapping to `format_args!` whatever happens is the best behavior, given how unlikely it is that somebody will call this macro with a constant string literal. (Calling it with a constant string literal wouldn't allow having the same type as another call with more than a constant string because the Fn type would be different.)

**Comment by Lucretiel** [comment]
Good question! So while it's true that people will never (or hardly ever) call `lazy_format!("Plain Unformatted String")`, `write` is also used in several other places. In particular it's used in the conditional version of `lazy_format!`:

```rust
lazy_format!(match (opt_int) {
    Some(x) => "{x}",
    None => "No value given"
})
```

This expands to something like:

```rust
make_lazy_format!(move |fmt| {
    match opt_int {
        Some(x) => lazy_format::write!(fmt, "{x}"),
        None => lazy_format::write!(fmt, "No value given"),
}
```

And from here the `None` branch expands to the shortcutted version: `::core::fmt::Write::write_str(&mut fmt, "No value given")`. The intent here is to skip around the `&dyn Display` that appears in essentially all calls to `format_args!`.

That being said, if there's ever a documented case of this `.any` not getting optimized away at compile time, I can replace this with a `const fn` version to force the check to happen at compile time.

**Comment by Lucretiel** [comment]
Addressed in 2.0.2; the check is now guaranteed to be computed as a `const`, to ensure it incurs no runtime penalty, rather than relying on an optimization.
